### DISCUSSION TITLE
Out-of-the-box Indentation Behavior

### DISCUSSION THREAD

**Discussion by spillz** [LazyVim/LazyVim#1798] [General]
Hi,

Just for background, I started using nvim recently to see if I can ween myself off VS code. Then I discovered lazyvim, which is very good. I have a long history of love-hate with VIM and am barely functional with it.

I'm a bit puzzled by what's going on with indentation/tab stops. I opened a 4-tab-space-indented javascript file, noticed it was showing as 2 spaces, ran `:set tabstop=4 shiftwidth=4` made some edits, then watched to my horror as :w converted everything to 2 spaces. Besides being puzzled that anyone would want behavior like this turned on by default, I was equally puzzled that Dr. Google tells me this is the behavior of the language server.  I regularly edit other people's code whose spacing preferences differ from mine but who am I to simply override their choice? Also puzzled that whatever features are enabled seemingly aren't able to detect indentation size automatically because all of my 4 space files were treated as 2 space files. And finally, why are these plugins, if they are indeed the cause, simply ignoring what the user sets in `tabstop` and `shiftwidth` with nary a warning?

Sorry, that was all a bit tongue in cheek but I am genuinely curious about what's going on here and the correct way to fix it to my liking.

**Comment by dpetka2001** [comment]
You can disable `format_on_save` with a keybinding `<leader>uf`. If you want to make the change permanent, you should put in your `lua/config/options.lua` the following `vim.g.autoformat = false`. The default Lazyvim configuration set `tabstop/shiftwidth` to 2. If you want to have different indentations for different filetypes (e.g lua, python, rust), you could make a file `after/ftplugin/python.lua` and put your options there. The directory `after` doesn't exist by default, so you should create it at the root of your Neovim configuration.

**Reply by luzik** [reply]
My options looks like that:
   7 ▎ local opt = vim.opt
   6 ▎ opt.confirm = false
   5 ▎ opt.expandtab = false
   4 ▎ opt.tabstop = 4
   3 ▎ opt.softtabstop = 0
   2 ▎ opt.shiftwidth = 4
   1 ▎ vim.g.python_recommended_style = 0
  11 ▎ vim.g.autoformat = false

but code format replace tabs with spaces. How to change this formatter indent style ? I use quite default lazyvim setup with only python extras enabled

**Reply by dpetka2001** [reply]
Are you using `black` for python formatting?

**Comment by luzik** [comment]
I belieave that default is ruff

**Reply by dpetka2001** [reply]
No, LazyVim by default doesn't use any formatter for Python. The user explicitly chooses what he wants. Did you enable from Extras the `formatting.black`?

**Comment by luzik** [comment]
LazyVim Extras shows only
lang.python  neotest-python  nvim-dap-python  nvim-lspconfig  nvim-treesitter  venv-selector.nvim


space-c-f is formatting my python files

do not know if it is related but :lspinfo show
Client: ruff_lsp
Client: pyright

**Comment by luzik** [comment]
and I was using https://github.com/LazyVim/starter

**Reply by dpetka2001** [reply]
Create a file `pyproject.toml` at the root of your Python project with contents
```toml
[tool.ruff.format]
indent-style = "tab"
```
Test it to see if you get tabs instead of spaces.

**Comment by luzik** [comment]
Yes it is working now. How can I put this into lazyvim configuration?

**Reply by dpetka2001** [reply]
This doesn't have anything to do with LazyVim configuration. As I said, LazyVim doesn't have a formatter for Python by default. The user chooses some external tool like `ruff` or `black` to do the formatting. Keep in mind that `black` doesn't support `<tab>` as indenting style (see [issue](https://github.com/psf/black/issues/47) here). Only `ruff` allows you to define `indent-style = "tab"`. 

Edit: also [pep8](https://peps.python.org/pep-0008/) suggests to use spaces instead of tabs. So keep that in mind when you decide to use tabs instead spaces. You might end up using a different convention than `pep8` suggestion if you want to contribute to some project. You should always take into account what the project uses.

**Comment by luzik** [comment]
I belieave that enabling lang.python extras installed ruff. I didn't done it by myself.

In old vim I was using
let g:ale_python_flake8_options = '--ignore=W191,E265,E501,E223,E117'
and was hope that can do something similar to inject 'indent-style = "tab"' to ruff from lazyvim config

**Reply by dpetka2001** [reply]
I don't know if you can pass format arguments to `ruff-lsp` via `lspconfig`. Your best bet for now would be using `pyproject.toml` at the root of your project or ask at the `ruff-lsp` repo if it's possible to achieve such a thing, as I don't know if it's possible.
